### DISCUSSION TITLE
Detect connection closed inside POST handler doing long-poll or SSE

### DISCUSSION THREAD

**Discussion by cameronelliott** [tokio-rs/axum#1094] [Q&A] [answered]
Thank you for writing Axum, I am amazed by it's power and comprehensiveness. 
(for example, I enabled Cors, and I was amazed how easy it was. I did it in Go before by hand, it was lots of work.)

I am writing a POST handler which is blocking on a future for an undetermined amount of time.
Kind of like long-polling or SSE.

I'd this handler to give-up and return if the originator goes away, ie: terminates the HTTP request or closes the TCP socket.

I am currently using ContentLengthLimit to get the request body.

I was thinking maybe I could use BodyStream or RawBody to detect a closed connection, 
_after_ I read the POST request body, but after thinking about it, I'm not sure that would work.

Is there a way I can pass a future into my handler somehow that I can use to detect a closed connection
somehow?

Thank you.

**Comment by davidpdrsn** [comment, marked as answer]
Happy to hear you're enjoying axum 😊

If I understand you correctly you don't actually need to do anything. It works automatically thanks to async rust and hyper. If the client closes the connection before the server sends a response, hyper will stop calling `poll` and drop the response future. That can be demonstrated by doing something like this:

```rust
use axum::{routing::get, Router};

#[tokio::main]
async fn main() {
    let app = Router::new().route("/", get(handler));

    axum::Server::bind(&"0.0.0.0:3000".parse().unwrap())
        .serve(app.into_make_service())
        .await
        .unwrap();
}

async fn handler() {
    // this will go out of scope either when this function returns (which it never does)
    // or hyper drops the future because the client closed the connection
    let _guard = Guard;

    // wait forever
    std::future::pending::<()>().await
}

struct Guard;

impl Drop for Guard {
    fn drop(&mut self) {
        println!("a `Guard` was dropped!")
    }
}
```

Try calling `GET /` in curl and then ctrl-c'ing it. You'll see the server prints "a `Guard` was dropped!" because `_guard` was dropped.

All this is unrelated to `ContentLengthLimit` which is for limiting the size of the request body.

**Reply by cameronelliott** [reply]
Duh, oh my gosh, thank you for being so nice.

I had added print statements after my 'await' to see if my future returned an error, but I forgot about the state-machine trickery
that allows the handler/parent-future to get cancelled without continuing execution.

Thanks for providing that example, and being so kind as to help me with something that's not specifically an Axum issue.

**Reply by davidpdrsn** [reply]
No worries :)
